JEP for Removal of Win32 (#15)
microsoft/openjdk-proposals: remove_win32_bit/jep.md

# JEP NNN: Removal of Windows 32-bit Support

Summary
-------

Following the deprecation approach in JEP 449, this proposal aims to take a further step by completely removing Windows 32-bit support from the OpenJDK. 
This decision acknowledges the evolving landscape of computing, where 64-bit systems have become predominant. 
The removal of 32-bit support will streamline development and focus resources on enhancing the performance and security of OpenJDK on modern architectures.

Goals
-----

- To cease all testing and development efforts targeting the Windows 32-bit platform within the OpenJDK project.
- To simplify OpenJDK's build and test infrastructure by focusing solely on Windows 64-bit support, aligning with current computing standards.

Non-Goals
---------

- This JEP does not aim to remove 32-bit support for platforms other than Windows.
- It does not intend to eliminate any 32-bit specific code paths that may exist that do not impact the Windows 32-bit platform.
- It does not intend to eliminate code on previous versions of OpenJDK.

Motivation
----------

> Building on the rationale of JEP 449, which deprecated the Windows 32-bit installer, this proposal recognizes the continued decline in the use and necessity of 32-bit Java applications on Windows.

Author: good eye

# JEP NNN: Removal of Windows 32-bit Support

_Author_	Bruno Borges & George Adams

> _Owner_ Bruno Borges

Reviewer: White space issues?

Author: I don't think these fields are really coming from the JEP text. I think they are JIRA. I am keeping here just to remember what to put on the JIRA fields.

_Type_	Feature
_Scope_	Implementation
_Status_	Draft
_Release_	23
_Component_	hotspot / other

> _Discussion_	jdk dash dev at openjdk dot org

Reviewer: Maybe porters-dev and build-dev as well if this field takes multiple lists?

Author: I don't think it does. We can check when we submit for real.

_Effort_	?
_Duration_	?
_Reviewed by_	Magnus Ihse Bursie
_Endorsed by_	Magnus Ihse Bursie
_Created_	<date>
_Updated_	<date>
_Issue_	<issue#>

Summary
-------

> Following the deprecation approach in JEP 449, this proposal aims to take a further step by completely removing Windows 32-bit support from the OpenJDK. 

Reviewer: ```suggestion
Following the deprecation of Windows 32-bit support in JEP 449, this proposal takes the next step by completely removing Windows 32-bit support from OpenJDK. 
```

This decision acknowledges the evolving landscape of computing, where 64-bit systems have become predominant. 

> The removal of 32-bit support will streamline development and focus resources on enhancing the performance and security of OpenJDK on modern architectures.

Reviewer: ```suggestion
The removal of Windows 32-bit support will streamline development and focus resources on enhancing the performance and security of OpenJDK on modern architectures.
```

Goals
-----

- To cease all testing and development efforts targeting the Windows 32-bit platform within the OpenJDK project.

> - To simplify OpenJDK's build and test infrastructure by focusing solely on Windows 64-bit support, aligning with current computing standards.

Reviewer 2: The complete removal of all code specific to just 32-bit Windows from the code base should be mentioned as a goal. This is kind of the most important one. There is no testing going on afaik, so saying that we are dropping it is not really saying anything. Nor are there any development efforts targeting 32-bit windows. 

The entire problem here is that there is a lot of legacy 32-bit code laying around which makes people hesitant as how to treat it when they come across it and need to change it.

> - To simplify OpenJDK's build and test infrastructure by focusing solely on Windows 64-bit support, aligning with current computing standards.

Reviewer 2: >  by focusing solely on Windows 64-bit support

I understand what you mean but it sounds like we're dropping support for all other platforms. :-) Can you please rephrase this so it is clear that it means that the Windows focus will be on 64-bit only?

Non-Goals
---------

- This JEP does not aim to remove 32-bit support for platforms other than Windows.

> - It does not intend to eliminate any 32-bit specific code paths that may exist that do not impact the Windows 32-bit platform.

Reviewer 2: I don't understand really how this is saying anything different from the first item, except making it sound more complicated? Can't you replace both of them with something like:

```
- This JEP does not aim to remove or change 32-bit support for any platforms other than Windows.
```

> - It does not intend to eliminate code on previous versions of OpenJDK.

Reviewer 2: "eliminate code or support for 32-bit Windows" to really drive in that point?

Motivation
----------

Building on the rationale of JEP 449, which deprecated the Windows 32-bit support, this proposal recognizes the continued decline in the use and necessity of 32-bit Java applications on Windows.
The computing industry has largely transitioned to 64-bit architectures, making the maintenance of a 32-bit JVM for Windows increasingly impractical and resource-intensive. 

> By removing this support, OpenJDK can allocate more resources towards innovations and improvements that benefit the broader community.

Reviewer 2: Something about simplifying the code base would be nice. As I said above, that is the main burden right of 32-bit Windows right now. (If you are testing 32-bit Windows in-house that is something I don't know about, but there is no publicly visible testing going on.)

1. **Build and Test Infrastructure**: Modifications to the OpenJDK build system to remove support for compiling on Windows 32-bit platforms, along with a halt in all testing activities for this architecture.
1. **Documentation and Communication**: Updates to OpenJDK documentation to reflect the removal of Windows 32-bit support and active communication within the community to ensure a smooth transition for users and developers.

Reference
---------

This proposal is inspired by and builds upon [JEP 449](https://openjdk.org/jeps/449), which deprecated the 32-bit Windows support from the OpenJDK.

Risks and Assumptions
---------------------

- **Risk**: There may be a subset of the user base that relies on 32-bit Java applications on Windows. Transition guidance and support by distributions and vendors of OpenJDK binaries will be critical.
- **Assumption**: Most current Java applications and users on Windows can migrate to 64-bit without significant obstacles, minimizing disruption.

Dependencies
------------

This JEP is independent of other proposals and is solely focused on the removal of 32-bit support for Windows within OpenJDK.

Impact
------

> - **Compatibility**: This change necessitates that applications running on 32-bit Java on Windows migrate to 64-bit Java or remain on legacy versions of OpenJDK that still include 32-bit support.

Reviewer 2: Maybe explain that "legacy" means "any version prior to JDK 23", since this change will only affect JDK 23 and onwards. (I guess in practice most users will remain on JDK8 as well, but that is not strictly necessary.)
